libsignal-service/src/cipher.rs

```diff
@@ -0,0 +1,210 @@
+                });
+            },
+        };
+
+        let mut plaintext = match envelope.r#type() {
+            Type::PrekeyBundle => {
+                let cipher =
+                    SessionCipher::new(&self.context, &self.store, &address)?;
+                let buf = cipher.decrypt_pre_key_message(
+                    &PreKeySignalMessage::deserialize(
+                        &self.context,
+                        ciphertext,
+                    )?,
+                )?;
+                Plaintext {
+                    metadata,
+                    data: Vec::from(buf.as_slice()),
+                }
+            },
+            Type::Ciphertext => {
+                let cipher =
+                    SessionCipher::new(&self.context, &self.store, &address)?;
+                let buf = cipher.decrypt_message(
+                    &SignalMessage::deserialize(&self.context, ciphertext)?,
+                )?;
+                Plaintext {
+                    metadata,
+                    data: Vec::from(buf.as_slice()),
+                }
+            },
+            Type::UnidentifiedSender => {
+                unimplemented!("UnidentifiedSender requires SealedSessionCipher, please report a bug against libsignal-service-rs.");
+            },
+            _ => {
+                unreachable!("conditions checked in previous match");
+            },
+        };
+
+        let version = self.store.load_session(&address)?.state().version();
+
+        strip_padding(version, &mut plaintext.data)?;
+        Ok(plaintext)
+    }
+
+    /// Equivalent of `SignalServiceCipher::getPreferredProtocolAddress`
+    fn get_preferred_protocol_address(
+        &self,
+        address: ServiceAddress,
+        device: u32,
+    ) -> Result<ProtocolAddress, ServiceError> {
+        let uuid = address
+            .uuid
+            .as_deref()
+            .map(|uuid| ProtocolAddress::new(uuid, device as i32));
+        let e164 = ProtocolAddress::new(&address.e164, device as i32);
+
+        if let Some(uuid) = uuid {
+            if self.store.contains_session(&uuid)? {
+                return Ok(uuid);
+            }
+        }
+
+        if self.store.contains_session(&e164)? {
+            return Ok(e164);
+        }
+
+        return Ok(ProtocolAddress::new(
+            address.get_identifier(),
+            device as i32,
+        ));
+    }
+}
+
+struct Plaintext {
+    metadata: Metadata,
+    data: Vec<u8>,
+}
+
+fn strip_padding(
+    version: u32,
+    contents: &mut Vec<u8>,
+) -> Result<(), ServiceError> {
+    if version < 2 {
+        return Err(ServiceError::InvalidFrameError {
+            reason: format!("Unknown version {}", version),
+        });
+    } else if version == 2 {
+        // No-op
+        return Ok(());
+    }
+
+    let mut start_of_padding = 0;
+    for (i, b) in contents.iter().enumerate().rev() {
+        if *b == 0x80 {
+            start_of_padding = i;
+            break;
+        }
+        if *b != 0x00 {
```

Comment:
Just the smallest nit to pick: this could still be `else if` though the previous `if` breaks the loop on a condition where `*b != 0x00`.

Reply:
Maybe best for being consistent with the Java version. I haven't really looking into *what* the loop does exactly, so I didn't really Rustify it decently... just mimiced the Java thing.

Reply:
I think it also helps readability, and though modern compilers probably know this stuff, I always like things explicit. Java has `else if` as well, though Java should also know how to compile that properly.